test/testvalueflow.cpp

```diff
@@ -3083,7 +3083,7 @@ class TestValueFlow : public TestFixture {
         values = tokenValues(code, "<");
         ASSERT_EQUALS(1, values.size());
         ASSERT(values.front().isPossible());
-        ASSERT_EQUALS(true, values.front().intvalue);
+        ASSERT_EQUALS(true, values.front().intvalue != 0);
```

Comment:
hmm this looks wrong. the result of a "<" can be 0 or 1. Your code implies that we get some other result?

Reply:
didn't this test that the intvalue was exactly 1 before?

Reply:
I must say.. I am skeptic about "bool => int" conversion warnings. This code does exactly what we intended.

Reply:
I think the problem is that it's not clear what the code is intending to do. I was subconsciously (and incorrectly) reading this as "is intvalue true" i.e. I was interpreting it as ```ASSERT_EQUALS((bool)true, (bool)values.front().intvalue)``` which I think is a reasonable interpretation (though incorrect in this case). The value ```(bool)values.front().intvalue``` is equivalent to ```values.front().intvalue != 0``` but the latter is more readable (by being more explicit) in my opinion.

The alternative (and correct in this case) interpretation is 
```ASSERT_EQUALS((long long)true, (long long)values.front().intvalue)``` 
but to me that raises the question: why not just  say ```ASSERT_EQUALS(1, values.front().intvalue)```?

Reply:
hmm.. imho, it's natural that it works like `(true == intvalue)` if we invent some other logic I'd feel that was a bug. If I see `(true == intvalue)` somewhere I would not think of that as strange or dangerous.

Imho `1` would be better than `(long long)true`. The `(long long)true` looks obfuscated. but well the result of a `<` is `true` I'd rather make the code explicit and let the compiler perform conversions and simplifications. I guess this is a stylistic decision and there is nothing wrong. :-(

If we want to replace `true` with `1` just here well it's not a huge deal. If we want to convert lots of code here and there .. then I am against it.

Can we try to disable this warning in VS solution so we don't get further tweaks.. do you use our solution or cmake?

Reply:
I use the VS solution.
The warning is: https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4805?view=msvc-160

It would be possible to suppress this warning either at the location where it occurs using a pragma or wholesale across the solution but I wouldn't recommend it.

Reply:
So can I get some more direction on this? As I see it the options are
1. compare explicitly as bools as I'm doing here
2. compare explicitly as ints like ```ASSERT_EQUALS(1, values.front().intvalue);```
3. suppress the warning here
4. suppress the warning everywhere
5. ignore the warning

My preference is for option 1 or 2. I'm not a fan of the idea of suppressing the warning everywhere as to me it seems like a valuable warning in general. From your comments thus far it seems like 2 makes more sense than 1. I'm happy to implement whichever solution you choose.

Reply:
> it seems like a valuable warning in general

I don't understand the idea of this warning. could you explain bugs it could avoid? What problems could you have with `true == x` where x is some arbitrary integer expression.. I think it seems obvious that it will work like `1 == x`?

I do feel that it could cause bugs to leave this warning on. With solution 1 that you suggest, the assertion does not work as it should.

Reply:
For clarity I have strike-throughed option 1.

The problem I imagine this causing is that it implies that the integer can only be true or false when that is not the case. If x!=true then to me that means x is logically false and thus ```bool b = x``` would be false. However if x is a value not equal to 0 or 1 then it is in a weird third state where it is neither true nor false and is treated as one or the other depending on the circumstances. Consider the weirdness of this code
```
int x = getx();
if (x)
 std::cout << "true"; //if x != 0
if (x==true)
  std::cout << "true"; //if x==1
```

Technically there's no bug but the code is needlessly ambiguous. Not to the compiler but to the reader. I think of this as being similar to the cppcheck warning that warns about needless ambiguity due to lack of parentheses.

It's interesting to note that msvc treats this as a level 1 warning implying that they consider it pretty important but gcc/clang seem to have no warning for it at any level.

Reply:
> Technically there's no bug but the code is needlessly ambiguous. Not to the compiler but to the reader. I think of this as being similar to the cppcheck warning that warns about needless ambiguity due to lack of parentheses.

I agree.. for the compiler it is 100% clear. it is not a technical question. the question is if cppcheck developers might misunderstand it. As you say it's the same thing with the parentheses warnings.

How many warnings do you see in Cppcheck repo?

Reply:
There's only one other which I believe to be unrelated to my changes. It's on line 388 of forwardanalyzer.cpp and reads
```} else if (condTok->values().front().intvalue == inElse) {``` where inElse is a bool.

Reply:
> There's only one other which I believe to be unrelated to my changes.

ok good does not seem to be a lot.

> It's on line 388 of forwardanalyzer.cpp and reads

I am not sure immediately what the intention is there.

Reply:
I've pushed option 2.